File: Level 7 - System Design & Probabilistic DS/vector_clocks/advanced_problems.cpp
#include <iostream>
#include <map>
#include <string>
#include <vector>
#include <chrono>

/**
 * Advanced Vector Clock Problems in C++
 * Focuses on Pruning strategy for bounded growth.
 */

class PrunableVC {
private:
    std::map<std::string, int> clock;
    std::map<std::string, long long> lastUpdate;
    int limit;

    long long now() {
        return std::chrono::system_clock::now().time_since_epoch().count();
    }

    void prune() {
        std::string oldest;
        long long minTime = -1;

        for (auto const& [node, time] : lastUpdate) {
            if (minTime == -1 || time < minTime) {
                minTime = time;
                oldest = node;
            }
        }
        if (!oldest.empty()) {
            std::cout << "[GC] Pruning node: " << oldest << std::endl;
            clock.erase(oldest);
            lastUpdate.erase(oldest);
        }
    }

public:
    PrunableVC(int l) : limit(l) {}

    void set(std::string node, int count) {
        clock[node] = count;
        lastUpdate[node] = now();
        if (clock.size() > limit) prune();
    }

    void print() {
        std::cout << "VC: { ";
        for (auto const& [k, v] : clock) std::cout << k << ":" << v << " ";
        std::cout << "} Size: " << clock.size() << std::endl;
    }
};

int main() {
    PrunableVC pvc(3);

    pvc.set("Node_1", 10);
    pvc.set("Node_2", 20);
    pvc.set("Node_3", 30);
    pvc.print();

    std::cout << "\nAdding Node_4..." << std::endl;
    pvc.set("Node_4", 1);
    pvc.print();

    return 0;
}

File: Level 7 - System Design & Probabilistic DS/vector_clocks/advanced_problems.java
import java.util.*;

/**
 * Advanced Vector Clock Problems in Java
 * Focuses on Pruning (Garbage Collection) to prevent unbounded growth.
 */
public class advanced_problems {

    static class PrunableVectorClock {
        Map<String, Integer> clock = new HashMap<>();
        Map<String, Long> lastUpdate = new HashMap<>(); // Physical time or logical sequence
        int maxNodes;

        PrunableVectorClock(int maxNodes) {
            this.maxNodes = maxNodes;
        }

        public void put(String nodeId, int count) {
            clock.put(nodeId, count);
            lastUpdate.put(nodeId, System.currentTimeMillis());

            if (clock.size() > maxNodes) {
                prune();
            }
        }

        /**
         * Pruning (Garbage Collection): Removes the oldest entry to keep size bounded.
         * Trade-off: Might lead to "False Independence" (missing a causality link).
         */
        private void prune() {
            String oldestNode = null;
            long oldestTime = Long.MAX_VALUE;

            for (Map.Entry<String, Long> entry : lastUpdate.entrySet()) {
                if (entry.getValue() < oldestTime) {
                    oldestTime = entry.getValue();
                    oldestNode = entry.getKey();
                }
            }

            if (oldestNode != null) {
                System.out.println("[Pruning] Removing node '" + oldestNode + "' to save space.");
                clock.remove(oldestNode);
                lastUpdate.remove(oldestNode);
            }
        }

        @Override
        public String toString() {
            return clock.toString() + " (Size: " + clock.size() + ")";
        }
    }

    public static void main(String[] args) throws InterruptedException {
        // Limit to 3 nodes
        PrunableVectorClock pvc = new PrunableVectorClock(3);

        System.out.println("Adding Node A, B, C...");
        pvc.put("A", 1);
        Thread.sleep(10);
        pvc.put("B", 5);
        Thread.sleep(10);
        pvc.put("C", 2);
        System.out.println("Current: " + pvc);

        System.out.println("\nAdding Node D (Should trigger pruning of A)...");
        pvc.put("D", 1);
        System.out.println("Current: " + pvc);

        System.out.println("\nUpdating Node B...");
        pvc.put("B", 6);

        System.out.println("\nAdding Node E (Should trigger pruning of C)...");
        pvc.put("E", 1);
        System.out.println("Final state: " + pvc);
    }
}
